### PR TITLE
Demand query cleanup

### PR DESCRIPTION
**Pull request by beyeler**
Following our discussion by email I cleaned the bidder demand query interface:

- remove all CPLEX parameters (Demand query will execute with its default parameters)
- add preMIP to MIPInstrumentation such that parameters for a Demand query can be changed by the user if needed

As I need the XORValueFunction to implement some new strategies I extended the DemandQuery functionallity there as well.

### CODE REVIEW

File: src/main/java/org/marketdesignresearch/mechlib/outcomerules/ccg/paymentrules/EuclideanNorm.java [hunk 1]
```diff
@@ -59,7 +59,6 @@ public PrimaryObjective getPrimaryObjectiveConstraint(IMIP program) {
         IMIP newProgram = MIPWrapper.makeMIPWithoutObjective(program);
         addNormObjective(newProgram);
         IMIPResult result = solveProgram(newProgram);
-        getMipInstrumentation().postMIP(MipInstrumentation.MipPurpose.PAYMENT, newProgram, result);
         metaInfo.setLpSolveTime(result.getSolveTime());
```
Comment (beyeler) [inline src/main/java/org/marketdesignresearch/mechlib/outcomerules/ccg/paymentrules/EuclideanNorm.java]: Please double check, but solveProgram already calls preMIP and postMIP so the postMIP here can be removed.

File: src/main/java/org/marketdesignresearch/mechlib/core/bidder/valuefunction/XORValueFunction.java [hunk 2]
```diff
@@ -48,7 +48,11 @@ public Bid toBid(UnaryOperator<BigDecimal> bundleBidOperator) {
     }
 
     public List<BundleValue> getOptimalBundleValueAt(Prices prices, int maxNumberOfBundles) {
-        return bundleValues.stream()
+        return this.getOptimalBundleValueAt(prices, maxNumberOfBundles, false);
+    }
```
Comment (beyeler) [inline src/main/java/org/marketdesignresearch/mechlib/core/bidder/valuefunction/XORValueFunction.java:52]: The return value of this function has changed. It does no longer return bundles with a negative utility. I did not find any reference where this function was used in a different way. Therefore, it is possible for me to change this.

File: src/main/java/org/marketdesignresearch/mechlib/core/bidder/valuefunction/XORValueFunction.java [hunk 3]
```diff
@@ -48,7 +48,11 @@ public Bid toBid(UnaryOperator<BigDecimal> bundleBidOperator) {
...
+    
+    public List<BundleValue> getOptimalBundleValueAt(Prices prices, int maxNumberOfBundles, boolean allowNegative) {
+        return bundleValues.stream().filter(b -> allowNegative || b.getAmount().subtract(prices.getPrice(b.getBundle()).getAmount()).signum() >= 0)
                 .sorted((a, b) -> {
                     BigDecimal first = a.getAmount().subtract(prices.getPrice(a.getBundle()).getAmount());
                     BigDecimal second = b.getAmount().subtract(prices.getPrice(b.getBundle()).getAmount());
```
Comment (beyeler) [inline src/main/java/org/marketdesignresearch/mechlib/core/bidder/valuefunction/XORValueFunction.java:58]: With this new function it is still possible to get bundles with a negative utility. Note that unlike the original filtering in the XORBidder it also returns bundles with a utility of 0 (which is not negative).

File: src/main/java/org/marketdesignresearch/mechlib/core/bidder/XORBidder.java [hunk 4]
```diff
@@ -61,9 +61,8 @@ public BigDecimal getValue(Bundle bundle) {
     }
 
     @Override
-    public List<Bundle> getBestBundles(Prices prices, int maxNumberOfBundles, boolean allowNegative, double relPoolTolerance, double absPoolTolerance, double poolTimeLimit) {
-        List<Bundle> result = value.getOptimalBundleValueAt(prices, maxNumberOfBundles).stream()
-                .filter(bundleValue -> allowNegative || bundleValue.getAmount().subtract(prices.getPrice(bundleValue.getBundle()).getAmount()).signum() > 0)
+    public List<Bundle> getBestBundles(Prices prices, int maxNumberOfBundles, boolean allowNegative) {
+        List<Bundle> result = value.getOptimalBundleValueAt(prices, maxNumberOfBundles, allowNegative).stream()
                 .map(BundleValue::getBundle).collect(Collectors.toList());
```
Comment (beyeler) [inline src/main/java/org/marketdesignresearch/mechlib/core/bidder/XORBidder.java:66]: Negative bundles are already filtered in the ValueFunction.

File: src/main/java/org/marketdesignresearch/mechlib/outcomerules/ccg/paymentrules/EuclideanNorm.java (hunk 1 above)
Comment (islerfab) [reply src/main/java/org/marketdesignresearch/mechlib/outcomerules/ccg/paymentrules/EuclideanNorm.java:62]: double checked, you're right 👍

**Pull request merged**